src/observers/slack_resource.cpp

```diff
@@ -23,39 +24,59 @@ Try<Nothing> SlackResourceObserver::consume(const ResourceUsage& usage) {
 
       auto previousSample = this->previousSamples->find(executor);
       if (previousSample != this->previousSamples->end()) {
-        Result<double_t> slackResource = CalculateCpuSlack(
-            (*previousSample), executor);
-        if (slackResource.isSome()) {
-            LOG(INFO) << NAME << "Estimated revocable resources for "
-                      << executor.executor_info().name() << " FrameworkID("
-                      << executor.executor_info().framework_id() << "): "
-                      << slackResource.get();
-            cpuSlack += slackResource.get();
-        } else if (slackResource.isError()) {
-          LOG(ERROR) << slackResource.error();
+        Result<double_t> executorCpuSlack = CalculateCpuSlack(
+            *previousSample, executor);
+        Try<double_t> executorCpuUsage = CountCpuUsage(
+            *previousSample, executor);
+        if (executorCpuSlack.isSome()) {
+          slackResources += executorCpuSlack.get();
+        } else if (executorCpuSlack.isError()) {
+          LOG(ERROR) << executorCpuSlack.error();
+        }
+
+        if (executorCpuUsage.isSome()) {
+          cpuUsage += executorCpuUsage.get();
+        } else if (executorCpuUsage.isError()) {
+          LOG(ERROR) << executorCpuSlack.error();
         }
       }
     }
   }
 
-  Resource slackResult;
-  Value_Scalar *cpuSlackScalar = new Value_Scalar();
-  cpuSlackScalar->set_value(cpuSlack);
+  Resources totalAgentResources(usage.total());
+  Option<double_t> totalAgentCpus = totalAgentResources.cpus();
+
+  if (totalAgentCpus.isSome()) {
+    const double_t maxSlack =
+        (maxOversubscriptionFraction * totalAgentCpus.get()) - cpuUsage;
+    if (maxSlack < slackResources) {
+      slackResources = maxSlack;
+    }
+
+    Resource slackResult;
+    Value_Scalar *cpuSlackScalar = new Value_Scalar();
+    cpuSlackScalar->set_value(slackResources);
 
-  slackResult.set_name("cpus");
-  slackResult.set_role(this->default_role);
-  slackResult.set_type(Value::SCALAR);
-  slackResult.set_allocated_scalar(cpuSlackScalar);
-  slackResult.set_allocated_revocable(new Resource_RevocableInfo());
+    slackResult.set_name("cpus");
+    slackResult.set_role(this->default_role);
+    slackResult.set_type(Value::SCALAR);
+    slackResult.set_allocated_scalar(cpuSlackScalar);
+    slackResult.set_allocated_revocable(new Resource_RevocableInfo());
 
-  Resources result(slackResult);
+    Resources result(slackResult);
 
-  produce(result);
+    produce(result);
 
-  this->previousSamples->clear();
-  this->previousSamples = std::move(newSamples);
+    this->previousSamples->clear();
+    this->previousSamples = std::move(newSamples);
 
-  return Nothing();
+    return Nothing();
+  } else {
```

Comment:
I would rather put that at the beginnig of consume.. Why doing all calculations when at the end slack observer realizes that it does not have totalAgentCpus at the first place? + It look's better when excluding the path than if deep branching e.g
```
if (totalAgentCpus.isNone())  {
  return Error(....);
}
```